From 63481710b6c6ce0f2978c79ff2fb3f22c3a54ed3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Mon, 23 Sep 2024 13:25:31 +0200
Subject: [PATCH 1407/1695] libutils: util.h: add usage description to
 {get,set}_field_u{32,64}()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Added comments to get_field_u{32,64}() and set_field_u{32,64}() to
improve understanding and reduce errors.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 199b93ee168eadf6835aa8764d6473006d74ec2d libutils: util.h: add usage description to {get,set}_field_u{32,64}()]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: If35adf1b3506fd865833bae3e88420292253e7ce
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/406204
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 lib/libutils/ext/include/util.h | 29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/lib/libutils/ext/include/util.h b/lib/libutils/ext/include/util.h
index ba83a673b..2384d87e5 100644
--- a/lib/libutils/ext/include/util.h
+++ b/lib/libutils/ext/include/util.h
@@ -187,7 +187,34 @@ static inline void reg_pair_from_64(uint64_t val, uint32_t *reg0,
 	*reg1 = low32_from_64(val);
 }
 
-/* Get and set bit fields  */
+/**
+ * @brief Functions to get and set bit fields in a 32/64-bit register.
+ *
+ * These inline functions allow setting and extracting specific bits in
+ * a register (`reg`) according to a given mask (`mask`). The mask
+ * specifies which bits in the register should be updated or extracted.
+ *
+ * - `set_field`: Modifies specific bits in a register by clearing the
+ *   bits specified by the mask and then setting these bits to the new
+ *   value (`val`).
+ * - `get_field`: Extracts the value of specific bits in a register by
+ *   isolating the bits specified by the mask and then shifting them to
+ *   the rightmost position.
+ *
+ * @param reg The original 32-bit or 64-bit register value.
+ * @param mask A bitmask indicating which bits in the register should be
+ * updated or extracted.
+ * @param val The new value to be inserted into the register at the
+ * positions specified by the mask (only for `set_field` functions).
+ * @return The updated register value with the specified bits set to the
+ * new value (for `set_field` functions), or the value of the bits
+ * specified by the mask, shifted to the rightmost position (for
+ * `get_field` functions).
+ *
+ * Note: These functions exist in both 32-bit (`set_field_u32`,
+ * `get_field_u32`) and 64-bit (`set_field_u64`, `get_field_u64`)
+ * versions.
+ */
 static inline uint32_t get_field_u32(uint32_t reg, uint32_t mask)
 {
 	return (reg & mask) / (mask & ~(mask - 1));
-- 
2.47.2

